ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java

```diff
@@ -0,0 +1,658 @@
+    List<Type> projectedFields = getProjectedGroupFields(schema, colNames, colTypes);
+    Type[] typesArray = projectedFields.toArray(new Type[0]);
+
+    return Types.buildMessage()
+      .addFields(typesArray)
+      .named(schema.getName());
+  }
+
+  /**
+   * Searchs column names by index on a given Parquet file schema, and returns its corresponded
+   * Parquet schema types.
+   *
+   * @param schema Message schema where to search for column names.
+   * @param colNames List of column names.
+   * @param colIndexes List of column indexes.
+   * @return A MessageType object of the column names found.
+   */
+  private static MessageType getSchemaByIndex(MessageType schema, List<String> colNames, List<Integer> colIndexes) {
+    List<Type> schemaTypes = new ArrayList<Type>();
+
+    for (Integer i : colIndexes) {
+      if (i < colNames.size()) {
+        if (i < schema.getFieldCount()) {
+          schemaTypes.add(schema.getType(i));
+        } else {
+          //prefixing with '_mask_' to ensure no conflict with named
+          //columns in the file schema
+          schemaTypes.add(Types.optional(PrimitiveType.PrimitiveTypeName.BINARY).named("_mask_" + colNames.get(i)));
+        }
+      }
+    }
+    return new MessageType(schema.getName(), schemaTypes);
+  }
+
+  /**
+   * columnBatch object that is used for batch decoding. This is created on first use and triggers
+   * batched decoding. It is not valid to interleave calls to the batched interface with the row
+   * by row RecordReader APIs.
+   * This is only enabled with additional flags for development. This is still a work in progress
+   * and currently unsupported cases will fail with potentially difficult to diagnose errors.
+   * This should be only turned on for development to work on this feature.
+   *
+   * When this is set, the code will branch early on in the RecordReader APIs. There is no shared
+   * code between the path that uses the MR decoders and the vectorized ones.
+   *
+   * TODOs:
+   *  - Implement v2 page formats (just make sure we create the correct decoders).
+   */
+  private VectorizedRowBatch columnarBatch;
+
+  @Override
+  public boolean next(
+    NullWritable nullWritable,
+    VectorizedRowBatch vectorizedRowBatch) throws IOException {
+    columnarBatch = vectorizedRowBatch;
+    return nextBatch();
+  }
+
+  /**
+   * Advances to the next batch of rows. Returns false if there are no more.
+   */
+  public boolean nextBatch() throws IOException {
+    initRowBatch();
+    columnarBatch.reset();
+    if (rowsReturned >= totalRowCount)
+      return false;
+    checkEndOfRowGroup();
+
+    int num = (int) Math.min(VectorizedRowBatch.DEFAULT_SIZE, totalCountLoadedSoFar - rowsReturned);
+    for (int i = 0; i < columnReaders.length; ++i) {
+      if (columnReaders[i] == null)
+        continue;
+      columnarBatch.cols[colsToInclude.get(i)].isRepeating = true;
```

Comment:
isRepeating is false by default. We set it to true at initialization and if one value is not the same as others, it will change to false.